Use non-default container port and scoped port vars in webhook e2e tests

jpbetz opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Conformance tests for CRD GA and Admission Webhooks GA should exercise the use of a non-default webhook port.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20180415-crds-to-ga.md)

/cc @liggitt @roycaihw @sttts @caesarxuchao 
/sig api-machinery
/area custom-resources

jpbetz commented:
Rebased on master, merged with v1 admission changes, e2e tests passed locally.

jpbetz commented:
/priority important-soon

jpbetz commented:
/retest

liggitt commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -601,8 +629,11 @@ func deployWebhookAndService(f *framework.Framework, image string, context *cert
 			},
 			Image: image,
+			// Use same port for service and containers.
+			Ports: []v1.ContainerPort{{ContainerPort: servicePort}},
```
it looks like we were already using a non-standard port for the service (which is good). I expected this PR to just change the container port to something that was not 443 and not the same as the service port (e.g. 9444). I would just hard code that into the pod and service spec.... I think that would make this PR a ~3 line change

jpbetz replied:
I've rewritten the PR to just swap the container port with a non-default one which is much simpler.  It's still more than 3 lines because I removed the global port variable and replaced it with a context scoped one which required threading a param through the stack, which is conceptually trivial but took quite a few lines to do.

jpbetz replied:
see latest PR update

liggitt commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -122,80 +124,88 @@ var _ = SIGDescribe("CustomResourceConversionWebhook", func() {
-					ConversionReviewVersions: []string{"v1", "v1beta1"},
-				},
+	ginkgo.Context("Running a conversion webhook service on a non-default port", func() {
+		servicePort = alternateServicePort
```
I lose track of which closures get run when. I thought variable assignments had to be done in beforeEach blocks, but maybe I'm mistaken

jpbetz replied:
Dropped the contexts in the latest PR update.

liggitt commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -122,80 +124,88 @@ var _ = SIGDescribe("CustomResourceConversionWebhook", func() {
-				},
+	ginkgo.Context("Running a conversion webhook service on the default port", func() {
+		ginkgo.BeforeEach(func() {
+			deployCustomResourceWebhookAndService(f, imageutils.GetE2EImage(imageutils.Agnhost), context, servicePort)
```
would `servicePort = defaultServicePort` make more sense here instead of the top-level BeforeEach?

jpbetz replied:
latest PR update removes this part too

liggitt commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -232,7 +242,7 @@ func createAuthReaderRoleBindingForCRDConversion(f *framework.Framework, namespa
 }
 
-func deployCustomResourceWebhookAndService(f *framework.Framework, image string, context *certContext) {
+func deployCustomResourceWebhookAndService(f *framework.Framework, image string, context *certContext, servicePort int32) {
```
same comment, pass distinct service and container ports

jpbetz committed:
Use non-default container port and scoped port vars in webhook e2e tests

jpbetz commented:
I've rewritten the PR to just swap the container port with a non-default one which is much simpler. It's still more than 3 lines because I removed the global port variable and replaced it with a context scoped one which required threading a param through the stack, which is conceptually trivial but took quite a few lines to do.

liggitt commented:
/lgtm
/approve

jpbetz commented:
/test pull-kubernetes-bazel-test
Build output shows that the tests passed (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/81541/pull-kubernetes-bazel-test/1163907284804308992) retrying
